graphql.go

```diff
@@ -9,25 +9,58 @@ import (
 	"net/http"
 
 	"github.com/shurcooL/graphql/internal/jsonutil"
-	"golang.org/x/net/context/ctxhttp"
 )
 
+var (
+	defaultClientHeaders = map[string]string{
+		"Content-Type": "application/json",
+	}
+)
+
+// ClientOptFunc graphql client option
+type ClientOptFunc func(*Client)
+
+// WithHeader set graphql client header
+func WithHeader(key, val string) ClientOptFunc {
+	return func(c *Client) {
+		c.headers[key] = val
+	}
+}
+
+// WithCookie set graphql client cookie
+func WithCookie(key, val string) ClientOptFunc {
+	return func(c *Client) {
+		if c.cookies == nil {
+			c.cookies = map[string]string{}
+		}
+		c.cookies[key] = val
+	}
+}
+
 // Client is a GraphQL client.
 type Client struct {
 	url        string // GraphQL server URL.
 	httpClient *http.Client
+
+	headers,
+	cookies map[string]string
 }
 
 // NewClient creates a GraphQL client targeting the specified GraphQL server URL.
 // If httpClient is nil, then http.DefaultClient is used.
-func NewClient(url string, httpClient *http.Client) *Client {
```

Comment:
Changing a function's signature is a very bad idea for backward compatibility. It's better to create a new function and have the old call the new. A blog entry from the Go team discusses this here: https://blog.golang.org/module-compatibility

Reply:
just add some option args, not change the function's behavior.

Reply:
Correct, you changed the function's *signature*. By adding args to the function, you broke backward compatibility. See the blog post from the Go team that I linked in the original message.  You should never change a function's signature in a module unless you're doing a major version release with breaking changes, and even then, try to avoid it.

Reply:
From the readme of the project: "Status: In active early research and development. The API will change when opportunities for improvement are discovered; it is not yet frozen.". 

I think breaking changes will be expected if they make sense.

Reply:
> you broke backward compatibility

it not broke the backward compatibility. I add **option args**, you can just ignore these args.

Reply:
The change here is to add a variadic, old implementations will not break. How is this a breaking change?

Reply:
Laisky, again, you changed the function's signature. Please see the provided reference on how it is a breaking change. The blog post from the Go team explains how it's a breaking change when you add args to an exported function.  From the provided link "Often, breaking changes come in the form of new arguments to a function." https://blog.golang.org/module-compatibility they also cover how to add arguments to an exported function without making it a breaking change. Don't change a function's signature if you don't need to..

Reply:
I add an new method called `NewClientWithOptions`